### ISSUE TITLE
CPU vs GPU Performance

### ISSUE THREAD

**Issue by Mazecreator** [tensorflow/tensorflow#3320]
I am working on a reinforcement learning model problem.  I have been working to get the model creation running faster and bumped into a strange issue I cannot explain.  It runs much faster ~50% or so on the CPU vs the GPU.  This was unexpected and I have disabled the GPU using **"export CUDA_VISIBLE_DEVICES=-1"** so the learning runs faster.  I have been looking at upgrading my GTX 950, but not sure it makes sense if I don't get a speed improvement.

I ran a profile based upon #1824 and got the following trace files for a single ".run()" iteration. I am not sure how to read this, but the GPU iteration took over 10ms where the CPU alone is <10ms. I am running the HEAD of TensorFlow (reports 0.9.0) on Ubuntu 15.10 with CUDA 7.5 and cuDNN 4 & 5 (tried both). The CPU is a dual XEON 6 core, 2.66 GHz  processors (24 threads total) with 72 GB or RAM (DDR3).

I have a GTX 950 GPU.  I can't tell if this performance difference is related to the structure of the graph or simply the data set isn't big enough to get a benefit from the GPU given the IO overhead?  I have tested the GPU with TF on a basic "matmut()" of a 7000x7000 matrix and it beats the CPU hands down by orders of magnitude.  So I known it is installed correctly.

Then networks runs in this case a Batch of 200 x 189 into 5 layers with Dropout() between each layer. The layers are 140, 120, 100, 80, and 3 as the output.  Any advice or things to try would be much appreciated.

**CPU Timeline:**
![cpu timeline](https://cloud.githubusercontent.com/assets/18412448/16854147/ad4c8760-49dd-11e6-8919-2c0940322b53.png)

**GPU Timeline:**
![gpu timeline](https://cloud.githubusercontent.com/assets/18412448/16854181/cfa25844-49dd-11e6-9f20-e2268bdd4299.png)

NOTE: This issue was posted to #2444 but may have got lost.  It might be related to RL performance but feel it is a separate issue.

I have enclosed the TIMELINE trace files for more detail.  If it is of value, I can create a "tensorboard" log file so you can review the model in detail.
[GPU Slowdown.zip](https://github.com/tensorflow/tensorflow/files/364706/GPU.Slowdown.zip)

**Comment by jmchen-g** [comment]
@zheng-xq Could you take a look at this please?

**Comment by zheng-xq** [comment]
A few observations: 

1. GTX 950 is definitely a low-end card for deep learning. It is roughly 1/4 of a Titan-X, which is a good card. 
2. You have a lot of small GPU kernels, which medium and large gaps in-between them. So the total amount of GPU compute time is probably much smaller than 10ms, but you GPU has a lot of idle time. 

A few possible things to try: 

A. Each of your kernel is too small, merge them into larger kernels. Use larger batch, and bigger shapes, so each op has more computation to it. For example, 7000 x 7000 matmul is something that is very good for GPU. 
B. Reduce the overhead of each op. Make sure your have as little memory transfer between devices as possible. Make sure you don't have other CPU intensive computation that holds the kernel launch. Also it is possible that you have too many threads launching GPU kernels that are thrashing each other. Play with inter_op_parallelism_threads to a smaller number. 

The best way to get more help is to upload a small repro case so more people can help. 

Have fun.

**Comment by Mazecreator** [comment]
@zheng-xq, sorry for the delay getting back to you but as I was putting the test case together I bumped into some unexpected results.

What I did was to take my **Original Solution** and stripped out the "Game Environment".  I replaced it with a random data generation.  In this Game Environment, there is no creation/modification of the TensorFlow Graph.  The structure closely follows/leverages [nivwusquorum's Github Reinforcement Learning Example](https://github.com/nivwusquorum/tensorflow-deepq/tree/master).

On 7/15/2016 I did a "git pull" to head for Tensorflow.  I executed the Graph with and without the GPU enabled and recorded the times (see attached chart).  The unexpected result is the GPU outperformed the CPU (which is the initial expectation that wasn't met).  So this code "cpuvsgpu.py" with the supporting libraries performs better with the GPU.  So I turned my attention to what may be different between my **Original Solution** and the published code.  I also update the head to 7/17/2016.  Something did improve as the overall difference between the CPU & GPU on the **Original Solution** is much closer than a week again where I was seeing 47s CPU vs 71s GPU .  A quick look at the new Traces vs my initial trace, seems like "summary's" may have been changed but there may have been other improvements as well. 

![gtx 950 timing](https://cloud.githubusercontent.com/assets/18412448/16904662/0cc86c72-4c68-11e6-9a79-c06bb1565215.png)

I tried 2 other combinations to better reflect how the **Original Solution** functioned.  Those were under heavy CPU load (~60% - 70%) and simulated that with concurrent execution of that script.  The other variation was to increase the "Data IO", the **Original Solution** uses lists of observations to randomly select observations for training.  This list has a fixed upper limit and then starts deleting the first item in the list while appending the new.  I figured maybe one of these was slowing down streaming of data to the GPU.  Unfortunately, neither of these version caused the CPU to outperform the GPU.  I also ran a quick GPUTESTER app which does large matrix multiplication to get a feel for timing differences with size of task and are as expected.

I have enclosed the test scripts as well as some of the raw data, Trace Files & TensorBoard log files to speed up any review.
[CPUvsGPU testing.zip](https://github.com/tensorflow/tensorflow/files/368246/CPUvsGPU.testing.zip)

I am going to keep looking at the **Original Solution** to see if I can figure out what is indirectly causing the GPU to perform poorly (any thoughts you have would be great).  I am going to start at the only real interface between the Game Environment and the graph and that is the "current_controller.store(observation, action, reward, new_observation)" line.  The only thing I can figure is maybe "observation"'s are not the same as what I think is generated by the Game Environment.  Again, not sure why GPU would perform poorly compared to CPU but would like to understand where this indirect affect on the graph is taking place.  Reinforcement Learning is slow as it is so any performance improvement would be appreciated.

I would like to follow-up with you on your comments from above as I really don't know how to improve this graph or where the small OPS may be hiding.  Also, batch size may affect convergence negatively but I can experiment with that future.  It seems like this is where most of the performance may be going.  Also, if you have any tricks to combine smaller ops into bigger ones without impacting the logic (function) of the graph.
